### ISSUE TITLE
[gfs] Compilation fails with USE_CPP14=OFF

### ISSUE THREAD

**Issue by krivenko** [TRIQS/triqs#360]
Multivar GFs are written using C++14 features. This part must be excluded from compilation if `USE_CPP14` is switched off.

I know, we are about to abolish the C++11 mode. Nonetheless, it must be functional as long as we still have it.

```
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:35:34: warning: 'decltype(auto)' type specifier is a C++14 extension [-Wc++14-extensions]
  template <typename T> decltype(auto) get_index(T &&x) { return std::forward<T>(x); }
                                 ^
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:35:25: error: deduced return types are a C++14 extension
  template <typename T> decltype(auto) get_index(T &&x) { return std::forward<T>(x); }
                        ^
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:36:25: error: 'auto' return without trailing return type; deduced return types are a C++14 extension
  template <typename T> auto get_index(mesh_point<T> const &x) { return x.index(); }
                        ^
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:37:25: error: 'auto' return without trailing return type; deduced return types are a C++14 extension
  template <typename T> auto get_index(mesh_point<T> &&x) { return x.index(); }
                        ^
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:38:25: error: 'auto' return without trailing return type; deduced return types are a C++14 extension
  template <typename T> auto get_index(mesh_point<T> &x) { return x.index(); }
                        ^
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:202:43: error: 'auto' return without trailing return type; deduced return types are a C++14 extension
  template <typename F, typename... Args> auto evaluate(interpol_t::Product, F const &f, Args &&... args) const {
                                          ^
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:92:33: error: 'auto' not allowed in lambda parameter
   return triqs::tuple::fold([](auto const &m, size_t R) { return R * m.size(); }, m_tuple, 1);
                                ^~~~
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:92:25: error: no member named 'fold' in namespace 'triqs::tuple'
   return triqs::tuple::fold([](auto const &m, size_t R) { return R * m.size(); }, m_tuple, 1);
          ~~~~~~~~~~~~~~^
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:98:27: error: 'auto' not allowed in lambda parameter
   auto l = [&res](int i, auto const &m) mutable { res[i] = m.size(); };
                          ^~~~
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:106:16: error: 'auto' not allowed in lambda parameter
   auto l = [](auto &p, auto const &m, auto const &i) { p = m.index_to_point(i); };
               ^~~~
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:106:25: error: 'auto' not allowed in lambda parameter
   auto l = [](auto &p, auto const &m, auto const &i) { p = m.index_to_point(i); };
                        ^~~~
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:106:40: error: 'auto' not allowed in lambda parameter
   auto l = [](auto &p, auto const &m, auto const &i) { p = m.index_to_point(i); };
                                       ^~~~
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:113:16: error: 'auto' not allowed in lambda parameter
   auto l = [](auto const &m, auto const &i) { return m.index_to_linear(i); };
               ^~~~
/home/igor/Physics/TRIQS/libs.git/triqs/./gfs/./meshes/product.hpp:113:31: error: 'auto' not allowed in lambda parameter
   auto l = [](auto const &m, auto const &i) { return m.index_to_linear(i); };
```

**Comment by krivenko** [comment]
I have to say that situation with `USE_CPP14` is a real mess.
Some source files contain preprocessor checks `#ifdef TRIQS_C11`, while the root `CMakeLists.txt` defines a different macro, `-DTRIQS_CPP11`.
Moreover, the GF library ignores `USE_CPP14` altogether.

**Comment by parcollet** [comment]
It is.
A simpler solution is to say that current master is take 1.4, it is C++14 only,
and then the dev is merged in master to become 1.5 later.
Do you really want to fix this ?

**Comment by krivenko** [comment]
>  A simpler solution is to say that current master is take 1.4, it is C++14 only,

Looks like the only viable option to me. I've been trying to add more workarounds...
C++14 is everywhere, in multi-variable and tensor-valued GFs, their wrappers, `triqs::utility::variant`, extractors. And then tests for all these parts.
